porch/pkg/git/draft.go

```diff
@@ -245,36 +246,110 @@ func (r *gitRepository) commitPackageToMain(ctx context.Context, d *gitPackageDr
 	}
 	packagePath := d.path
-	packageTree := d.tree
+	packageRevision := d.revision
+
+	// Fetch the commits belonging to this package revision.
+	commits, err := r.loadPackageCommits(d.commit, packagePath, packageRevision)
+	if err != nil {
+		return zero, zero, nil, fmt.Errorf("failed to load commits for package %s: %w", packagePath, err)
+	}
+	reverseCommitSlice(commits)
 
 	// TODO: Check for out-of-band update of the package in main branch
 	// (compare package tree in target branch and common base)
-	ch, err := newCommitHelper(repo, r.userInfoProvider, headCommit.Hash, packagePath, packageTree)
+	var ch *commitHelper
+	if len(commits) == 0 {
+		// If we can't find any commits, the draft might have been created outside of porch. We
+		// just add the changes to the main branch in a single commit.
+		ch, err = newCommitHelper(repo, r.userInfoProvider, headCommit.Hash, packagePath, d.tree)
+		if err != nil {
+			return zero, zero, nil, fmt.Errorf("failed to initialize commit of package %s to %s", packagePath, localRef)
+		}
+	} else {
+		// If we have commits, we reproduce the same commits on the main branch to keep
+		// the history from the draft.
+		ch, err = newCommitHelper(repo, r.userInfoProvider, headCommit.Hash, packagePath, plumbing.ZeroHash)
+		if err != nil {
+			return zero, zero, nil, fmt.Errorf("failed to initialize commit of package %s to %s", packagePath, localRef)
+		}
+
+		for _, commit := range commits {
+			// Look up the tree for the package in the commit.
+			t, err := commit.Tree()
+			if err != nil {
+				return zero, zero, nil, fmt.Errorf("cannot resolve commit's (%s) tree (%s) to tree object: %w", t.Hash, commit.TreeHash, err)
+			}
+			te, err := t.FindEntry(packagePath)
+			if err != nil {
+				return zero, zero, nil, err
+			}
+
+			err = ch.storeTree(packagePath, te.Hash)
+			if err != nil {
+				return zero, zero, nil, err
+			}
+
+			_, _, err = ch.commit(ctx, commit.Message, packagePath)
+			if err != nil {
+				return zero, zero, nil, fmt.Errorf("failed to commit package %s to %s", packagePath, localRef)
+			}
+		}
+	}
+
+	// Add a commit without changes to mark that the package revision is approved.
+	message := fmt.Sprintf("Approve %s/%s", packagePath, d.revision)
+	commitHash, newPackageTreeHash, err = ch.commit(ctx, message, packagePath)
 	if err != nil {
-		return zero, zero, nil, fmt.Errorf("failed to initialize commit of package %s to %s", packagePath, localRef)
+		return zero, zero, nil, fmt.Errorf("failed to commit package %s to %s", packagePath, localRef)
 	}
 
-	// Add the tasks to the commit message.
-	var tasks []*v1alpha1.Task
-	for i := range d.tasks {
-		t := d.tasks[i]
-		tasks = append(tasks, &t)
+	return commitHash, newPackageTreeHash, localTarget, nil
+}
+
+// TODO: This is an almost direct copy of
+// https://github.com/GoogleContainerTools/kpt/blob/3c3288af0c4c4a7e07ffeb6fe473d32afd81137b/porch/pkg/git/git.go#L860.
+// Fix it when we can use generics.
```

Comment:
We could just start using a comment like `// TODO:generics` or something like that